support/types.go

```diff
@@ -0,0 +1,100 @@
+	// using the cluster provider native way
+	GetKubeconfig() string
+
+	// GetKubeCtlContext is used to extract the kubectl context to be used while performing the operation
+	GetKubeCtlContext() string
+
+	// ExportLogs is used to export the cluster logs via the cluster provider native workflow. This
+	// can be used to export logs from the cluster after test failures for example to analyze the test
+	// failures better after the fact.
+	ExportLogs(dest string) error
+
+	// Destroy is used to cleanup a cluster brought up as part of the test workflow
+	Destroy() error
+
+	// LoadDockerImage is used to load a set of Docker images to the cluster via the cluster provider native workflow
+	// Not every provider will have a mechanism like this/need to do this. So, providers that do not have this support
+	// can just provide a no-op implementation to be compliant with the interface
+	LoadDockerImage(image string) error
+
+	// LoadImageArchive is used to provide a mechanism where a tar.gz archive containing the docker images used
+	// by the services running on the cluster can be imported and loaded into the cluster prior to the execution of
+	// test if required.
+	// Not every provider will have a mechanism like this/need to do this. So, providers that do not have this support
+	// can just provide a no-op implementation to be compliant with the interface
+	LoadImageArchive(archivePath string) error
+
+	// SetDefaults is a handler function invoked after creating an object of type E2EClusterProvider. This method is
+	// invoked as the first step after creating an object in order to make sure the default values for required
+	// attributes are setup accordingly if any.
+	// TODO: Find a better way to do this with the generics. We only need this because of Generics right now
+	SetDefaults() E2EClusterProvider
+
+	// WaitForControlPlane is a helper function that can be used to indiate the Provider based cluster create workflow
+	// that the control plane is fully up and running. This method is invoked after the Create/CreateWithConfig handlers
+	// and is expected to return an error if the control plane doesn't stabilize. If the provider being implemented
+	// does not have a clear mechanism to identify the Control plane readiness or is not required to wait for the control
+	// plane to be ready, such providers can simply add a no-op workflow for this function call.
+	// Returning an error message from this handler will stop the workflow of e2e-framework as returning an error from this
+	// is considered as  failure to provision a cluster
+	WaitForControlPlane(klient.Client) error
+
+	// KubernetesClient is a helper function used to provide an instance of kubernetes.Interface which can then be used
+	// to interact with Provider clusters in a generic way if the existing resources handlers/helpers are not enough
+	KubernetesClient() (kubernetes.Interface, error)
```

Comment:
Maybe we also need a `dynamic.Interface`, `*rest.Config` and `*rest.RESTClient`

Reply:
Hmm. Wondering if it would be a better option to just provide `rest.Config` from here and let the consumers build the client for their liking? That will also keep the external facing interfaces limited.